Source/CesiumRuntime/Private/CesiumGeoreferenceComponent.cpp

```diff
@@ -135,90 +138,293 @@ void UCesiumGeoreferenceComponent::InaccurateMoveToECEF(
+    UE_LOG(
+        LogCesium,
+        Verbose,
+        TEXT(
+            "Attaching CesiumGeoreferenceComponent callback to Georeference %s"),
+        *this->Georeference->GetFullName());
+    this->Georeference->OnGeoreferenceUpdated.AddUniqueDynamic(
+        this,
+        &UCesiumGeoreferenceComponent::HandleGeoreferenceUpdated);
+  }
+}
+
+void UCesiumGeoreferenceComponent::ApplyWorldOffset(
+    const FVector& InOffset,
+    bool bWorldShift) {
+  UE_LOG(
+      LogCesium,
+      Verbose,
+      TEXT("Called ApplyWorldOffset on component %s"),
+      *this->GetName());
+  UActorComponent::ApplyWorldOffset(InOffset, bWorldShift);
+  if (!this->FixTransformOnOriginRebase) {
+    return;
+  }
+
+  const UWorld* world = this->GetWorld();
+  if (!IsValid(world)) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT("CesiumGeoreferenceComponent %s is not spawned in world"),
+        *this->GetName());
+    return;
+  }
+  if (!IsValid(this->Georeference)) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT(
+            "CesiumGeoreferenceComponent %s does not have a valid Georeference"),
+        *this->GetName());
+    return;
+  }
+
+  // Compute the position that the world origin will have
+  // after the rebase, indeed by SUBTRACTING the offset
+  const glm::dvec3 oldWorldOriginLocation =
+      VecMath::createVector3D(world->OriginLocation);
+  const glm::dvec3 offset = VecMath::createVector3D(InOffset);
+  const glm::dvec3 newWorldOriginLocation = oldWorldOriginLocation - offset;
+
+  // Compute the absolute location based on the ECEF
+  // Do NOT use TransformEcefToUnreal, because it will return the
+  // position relative to the current world origin!
+  const glm::dmat4& ecefToUnreal =
+      this->Georeference->getGeoTransforms()
+          .GetEllipsoidCenteredToUnrealWorldTransform();
+  const glm::dvec3 absoluteLocation =
+      ecefToUnreal * glm::dvec4(_currentEcef, 1.0);
+
+  // Compute the new (high-precision) relative location from
+  // the absolute location and the new world origin
+  const glm::dvec3 newRelativeLocation =
+      absoluteLocation - newWorldOriginLocation;
+
+  const glm::dmat3 actorRotation = _getRotationFromActor();
+  _updateActorTransform(actorRotation, newRelativeLocation);
+}
 #if WITH_EDITOR
+void UCesiumGeoreferenceComponent::PreEditChange(
+    FProperty* PropertyThatWillChange) {
+  Super::PreEditChange(PropertyThatWillChange);
+
+  UE_LOG(
+      LogCesium,
+      Verbose,
+      TEXT("Called PreEditChange for %s"),
+      *this->GetName());
+
+  // If the Georeference is modified, detach the `HandleGeoreferenceUpdated`
+  // callback from the current instance
+  FName propertyName = PropertyThatWillChange->GetFName();
+  if (propertyName ==
+      GET_MEMBER_NAME_CHECKED(UCesiumGeoreferenceComponent, Georeference)) {
+    if (IsValid(this->Georeference)) {
+      this->Georeference->OnGeoreferenceUpdated.RemoveAll(this);
+      _updateActorTransform();
```

Comment:
Assuming that `this->Georeference` is pointing to the old Georeference pre-change, should we wait to call `_updateActorTransform` until `PostEditChangeProperty`?

Reply:
The intention behind calling `_updateActorTransform` was to ~"have an up-to-date state, _even if_ the Georeference is set to `None`/`null`". In an ideal world, this should not be necessary, because the state should be ~"up-to-date _at any point in time, anyhow_", but ... some of the Unreal lifecycle method calls are ... surprising and unexpected. (E.g. it does call `OnUnregister` on the component when editing a property - why?!?!). 

It might be possible to remove that `_updateActorTransform` call there, but I'd dedicatedly check whether it might lead to an inconsistent state when setting it to `None`. (Maybe even cover that with a unit test later...)

Reply:
@javagl That makes sense, thanks for clarifying. Perhaps it may be possible to remove, but this can be a question for later. For now, a simple comment on that line would be great.

Reply:
I have added a comment accordingly. 

The "broader" question (that we occasionally talked about, but where a final, definite answer is difficult) is: Which class should be able to cope with a `None`/`null` Georeference, and _how exactly should it behave in this case_?

Right now, the GeoRefComponent does lots of `IsValid` checks, and usually tries to handle the `else`-case with "default values" (like identity matrices). But I'd hesitate to establish an strong guarantee along the lines of "This works and makes sense in all cases".